File: c/queen_attack.c
#include "queen_attack.h"

#include <stdbool.h>
#include <stdlib.h>

bool is_valid(uint8_t n);
bool is_valid_position(position_t p);

attack_status_t can_attack(position_t queen_1, position_t queen_2) {
  if (!is_valid_position(queen_1)) {
    return INVALID_POSITION;
  }

  if (!is_valid_position(queen_2)) {
    return INVALID_POSITION;
  }

  // same position
  if (queen_1.row == queen_2.row && queen_1.column == queen_2.column) {
    return INVALID_POSITION;
  }

  // same row or same column
  if (queen_1.row == queen_2.row || queen_1.column == queen_2.column) {
    return CAN_ATTACK;
  }

  // diagonal
  int dx = abs(queen_1.row - queen_2.row);
  int dy = abs(queen_1.column - queen_2.column);
  if (dx == dy) {
    return CAN_ATTACK;
  }

  return CAN_NOT_ATTACK;
}

bool is_valid(uint8_t n) { return n < 8; }

bool is_valid_position(position_t p) {
  return is_valid(p.row) && is_valid(p.column);
}

File: python/queen_attack.py
from typing import Self


class Queen:

    def __init__(self, row: int, column: int):
        if row < 0:
            raise ValueError('row not positive')

        if row >= 8:
            raise ValueError('row not on board')

        if column < 0:
            raise ValueError('column not positive')

        if column >= 8:
            raise ValueError('column not on board')

        self.row = row
        self.column = column

    def can_attack(self, another_queen: Self) -> bool:
        if self.row == another_queen.row and self.column == another_queen.column:
            raise ValueError(
                'Invalid queen position: both queens in the same square')

        return self.row == another_queen.row or self.column == another_queen.column or abs(
            self.row - another_queen.row) == abs(self.column -
                                                 another_queen.column)
